### PR TITLE
Midi 37: random forest

### PR DESCRIPTION
**Pull request by SamuelJanas**
[MIDI-37](https://sacrebleu.atlassian.net/browse/MIDI-37)

[MIDI-37]: https://sacrebleu.atlassian.net/browse/MIDI-37?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by SamuelJanas** [comment]
Created a MNIST clone with embeddings of size 32 ([hugging face link](https://huggingface.co/datasets/SneakyInsect/MNIST-preprocessed)).
## Added a visualization for embeddings:
Since the embedding can't be displayed in 2D because of the size (between 5x5 and 6x6). I've added some padding(4 trailing black pixels in this case)
reconstructions_AE_MNIST

EDIT: Thought that it would be a good idea to add columns with different embedding sizes for even more convenience

**Comment by SamuelJanas** [comment]
## Summary of Results:

- Embedding size 32: 10.44% accuracy
- Embedding size 16: 10.83% accuracy
- Embedding size 8: 95.86% accuracy
- Re-test on embedding size 32 model: 96.23% accuracy
## Observations:
The accuracy results for embedding sizes 32 and 16 are close to random. This happened due to an issue when changing the dataset structure, causing order randomization. I'm currently working on a way to fix it.

The high accuracy (95.86%) for embedding size 8 suggests that the encoding process does provide valuable information. This is further confirmed by the re-tested accuracy of 96.23% for the embedding size 32 model (Use `test.py` to check the accuracy on any checkpoint).

**Comment by SamuelJanas** [comment]
Seeking feedback even though this is still a work in progress. Perhaps there's a straightforward solution that I'm overlooking. 😅

File: train_random_tree.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from datasets import load_dataset
+from sklearn.metrics import accuracy_score
```
Comment (roszcz) [inline train_random_tree.py:2]: `echo 'scikit-learn' >> requirements.txt` :)

File: train_random_tree.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+if __name__ == "__main__":
+    train_embeddings, train_labels, test_embeddings, test_labels = load_preprocessed_dataset("SneakyInsect/MNIST-preprocessed")
+
+    clf = RandomForestClassifier(n_estimators=100)
+    clf.fit(train_embeddings, train_labels)
+    # Predict the labels of the test data
+    predictions = clf.predict(test_embeddings)
+    # Calculate the accuracy of the predictions
+    accuracy = accuracy_score(test_labels, predictions)
+    print(f"Accuracy of the Random Forest classifier with embedding size {32}: {accuracy*100:.2f}%")
```
Comment (roszcz) [inline train_random_tree.py:28]:
```suggestion
def main() -> RandomForestClassifier:
    train_embeddings, train_labels, test_embeddings, test_labels = load_preprocessed_dataset("SneakyInsect/MNIST-preprocessed")

    clf = RandomForestClassifier(n_estimators=100)
    clf.fit(train_embeddings, train_labels)
    # Predict the labels of the test data
    predictions = clf.predict(test_embeddings)
    # Calculate the accuracy of the predictions
    accuracy = accuracy_score(test_labels, predictions)
    print(f"Accuracy of the Random Forest classifier with embedding size {32}: {accuracy*100:.2f}%")
    return clf
    
if __name__ == "__main__":
  main()
```

I think it's better to keep the `__main__` block to minimum

**Comment by SamuelJanas** [comment]
Reverted hugginface dataset to embedding size 32 only, it feels enough for this experiment given previous results.
I also tested some other embedding sizes, here are their reconstructions and random forest classifier accuracy:
### embedding size 4:
- Accuracy: 94.27%
reconstructions_AE_MNIST

### embedding size 2:
- Accuracy: 76.84%
emb_size_2

### embedding size 1: 🤔 
- Accuracy: 51.89%
reconstructions_AE_MNIST

**Comment by roszcz** [comment]
I think it's not necessary to convert embeddings into images - maybe doing `plt.plot(embedding)` instead could also give you some intuitions about what's going on 🤔

**Comment by SamuelJanas** [comment]
> I think it's not necessary to convert embeddings into images - maybe doing `plt.plot(embedding)` instead could also give you some intuitions about what's going on 🤔

Here's the updated version, decided on histogram, easier to look at distribution.
embedding histgram

Color of the histogram ~~can~~ should be changed to something more humane but the decision would take too long for me

**Review by roszcz — APPROVED** [review]
